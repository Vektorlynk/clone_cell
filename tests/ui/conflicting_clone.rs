use std::rc::Rc;

use clone_cell::{cell::Cell, clone::PureClone};

#[derive(PureClone)]
struct Foo {
    data: i32,
    ptr: Rc<Cell<Option<Foo>>>,
}

impl Clone for Foo {
    // Bad clone implementation. Example from:
    // https://users.rust-lang.org/t/why-does-cell-require-copy-instead-of-clone/5769/9
    fn clone(&self) -> Self {
        let data = &self.data;
        // Clear out the cell we're contained in...
        self.ptr.set(None);
        Self { data: self.data, ptr: self.ptr.clone() }
    }
}

fn main() {
    let c = Rc::new(Cell::new(None));
    c.set(Some(Foo {
        data: 42,
        ptr: c.clone()
    }));
    assert_eq!(c.get().unwrap().data, 42);
}
