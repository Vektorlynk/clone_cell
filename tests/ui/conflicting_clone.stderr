error[E0119]: conflicting implementations of trait `std::clone::Clone` for type `Foo`
  --> $DIR/conflicting_clone.rs:5:10
   |
5  | #[derive(PureClone)]
   |          ^^^^^^^^^ conflicting implementation for `Foo`
...
11 | impl Clone for Foo {
   | ------------------ first implementation here
   |
   = note: this error originates in the derive macro `PureClone` (in Nightly builds, run with -Z macro-backtrace for more info)
